File: libraries/kafka-client-py/kafka_client_py/headers.py
from collections.abc import Iterable
from datetime import UTC, datetime

KafkaHeaders = list[tuple[str, bytes]]


HEADER_ORIGINAL_TOPIC = "x-original-topic"
HEADER_ORIGINAL_PARTITION = "x-original-partition"
HEADER_ORIGINAL_OFFSET = "x-original-offset"

HEADER_RETRY_COUNT = "x-retry-count"
HEADER_RETRY_AT = "x-retry-at"
HEADER_FIRST_FAILED_AT = "x-first-failed-at"

HEADER_ERROR_CLASS = "x-error-class"
HEADER_ERROR_MESSAGE = "x-error-message"
HEADER_ERROR_STACK = "x-error-stack"
HEADER_FAILED_AT = "x-failed-at"

HEADER_CORRELATION_ID = "x-correlation-id"


def encode(value: str | int | datetime) -> bytes:
    if isinstance(value, datetime):
        if value.tzinfo is None:
            value = value.replace(tzinfo=UTC)
        return value.astimezone(UTC).isoformat().encode("utf-8")

    return str(value).encode("utf-8")


def decode(value: bytes | None) -> str | None:
    if value is None:
        return None
    return value.decode("utf-8")


def get(
    headers: Iterable[tuple[str, bytes]] | None,
    name: str,
) -> str | None:
    if not headers:
        return None
    last_match_bytes: bytes | None = None
    for header_name, header_value in headers:
        if header_name == name:
            last_match_bytes = header_value
    return decode(last_match_bytes)


def get_int(
    headers: Iterable[tuple[str, bytes]] | None,
    name: str,
    default: int = 0,
) -> int:
    raw_header_value = get(headers, name)
    if raw_header_value is None:
        return default

    try:
        return int(raw_header_value)
    except ValueError:
        return default


def get_datetime(
    headers: Iterable[tuple[str, bytes]] | None,
    name: str,
) -> datetime | None:
    raw_header_value = get(headers, name)
    if raw_header_value is None:
        return None

    try:
        return datetime.fromisoformat(raw_header_value)
    except ValueError:
        return None


def merge(
    base: KafkaHeaders | None,
    *additions: tuple[str, str | int | datetime],
) -> KafkaHeaders:
    merged_headers: KafkaHeaders = list(base) if base else []
    for name, value in additions:
        merged_headers.append((name, encode(value)))

    return merged_headers

File: libraries/kafka-client-go/headers/headers.go
package headers

import (
	"strconv"
	"time"
)

const (
	OriginalTopic     = "x-original-topic"
	OriginalPartition = "x-original-partition"
	OriginalOffset    = "x-original-offset"

	RetryCount    = "x-retry-count"
	RetryAt       = "x-retry-at"
	FirstFailedAt = "x-first-failed-at"

	ErrorClass   = "x-error-class"
	ErrorMessage = "x-error-message"
	ErrorStack   = "x-error-stack"
	FailedAt     = "x-failed-at"

	CorrelationID = "x-correlation-id"
)

type Header struct {
	Key   string
	Value []byte
}

type KafkaHeaders []Header

func String(key string, value string) Header {
	return Header{Key: key, Value: []byte(value)}
}

func Int(key string, value int64) Header {
	return Header{Key: key, Value: []byte(strconv.FormatInt(value, 10))}
}

func Time(key string, value time.Time) Header {
	return Header{Key: key, Value: []byte(EncodeTime(value))}
}

func EncodeTime(value time.Time) string {
	return value.UTC().Format(time.RFC3339Nano)
}

func Get(headers KafkaHeaders, name string) (string, bool) {
	var lastMatch []byte
	isFound := false
	for _, header := range headers {
		if header.Key == name {
			lastMatch = header.Value
			isFound = true
		}
	}

	if !isFound {
		return "", false
	}

	return string(lastMatch), true
}

func GetInt(headers KafkaHeaders, name string, defaultValue int) int {
	rawHeaderValue, isFound := Get(headers, name)
	if !isFound {
		return defaultValue
	}

	parsedValue, parseErr := strconv.Atoi(rawHeaderValue)
	if parseErr != nil {
		return defaultValue
	}

	return parsedValue
}

func GetTime(headers KafkaHeaders, name string) (time.Time, bool) {
	rawHeaderValue, isFound := Get(headers, name)
	if !isFound {
		return time.Time{}, false
	}

	parsedTime, parseErr := time.Parse(time.RFC3339Nano, rawHeaderValue)
	if parseErr != nil {
		return time.Time{}, false
	}

	return parsedTime, true
}

func Merge(base KafkaHeaders, additions ...Header) KafkaHeaders {
	mergedHeaders := make(KafkaHeaders, 0, len(base)+len(additions))
	mergedHeaders = append(mergedHeaders, base...)
	mergedHeaders = append(mergedHeaders, additions...)

	return mergedHeaders
}
